plat/mediatek/mt8183/drivers/gpio/mtgpio.c

```diff
@@ -0,0 +1,370 @@
+/*
+ * Copyright (c) 2018, ARM Limited and Contributors. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <assert.h>
+#include <debug.h>
+#include <mmio.h>
+#include <platform_def.h>
+#include <mcucfg.h>
+#include <spm.h>
+#include <delay_timer.h>
+#include <gpio/mtgpio.h>
+#include <gpio/mtgpio_cfg.h>
+#include <gpio.h>
+
+/******************************************************************************
+ MACRO Definition
+******************************************************************************/
+//#define  GIO_SLFTEST
+#define VERSION     GPIO_DEVICE
+/*---------------------------------------------------------------------------*/
+#define REGSET (4)
+#define REGCLR (8)
+#define REGMWR (0xC)
+#define GPIO_MODE_BITS                 4
+#define MAX_GPIO_MODE_PER_REG          8
+#define MAX_GPIO_REG_BITS              32
+/*---------------------------------------------------------------------------*/
+#define TRUE                            1
+#define FALSE                           0
+
+#define GIO_INVALID_OBJ(ptr)            ((ptr) != gpio_obj)
+/******************************************************************************
+Enumeration/Structure
+******************************************************************************/
+/*---------------------------------------------------------------------------*/
+void mt_set_gpio_dir_chip(uint32_t pin, int dir)
+{
+	uint32_t bit;
+
+	assert(pin < MAX_GPIO_PIN);
+
+	assert(dir < GPIO_DIR_MAX);
+
+	bit = pin % MAX_GPIO_REG_BITS;
+
+	/* To reduce memory usage, we don't use DIR_addr[] array*/
+	if (dir == GPIO_DIR_IN)
+		mmio_write_32(DIR_addr[pin].addr + 0x8, 1U << bit);
```

Comment:
I am very confused about these arrays (like DIR_addr)... where do they come from? I don't see them defined anywhere.

Reply:
arm-trusted-firmware\plat\mediatek\mt8183\drivers\gpio\mtgpio_cfg.h

Reply:
Ah... don't know why I overlooked so much stuff here last time.

Those arrays are going to waste a huge amount of memory and binary size. Isn't there some more programmatic way you could implement this? It seems like most of the registers (mode, out, in) are completely 100% regular according to a single GPIO base, so you should be able to get the addresses and bit positions with a single division. For the pull it looks like you just have a handful of consecutive regions each mapped to their own base address, so can't you just make a gpio_get_iobase() function that has a big switch statement with ranges or something?